components/brave_wallet/browser/asset_discovery_manager.cc

```diff
@@ -330,6 +365,248 @@ void AssetDiscoveryManager::MergeDiscoveredEthAssets(
+  if (!url.is_valid()) {
+    std::move(callback).Run({});
+    return;
+  }
+  std::vector<mojom::BlockchainTokenPtr> nfts_so_far = {};
+  auto internal_callback =
+      base::BindOnce(&AssetDiscoveryManager::OnFetchNFTsFromSimpleHash,
+                     weak_ptr_factory_.GetWeakPtr(), std::move(nfts_so_far),
+                     coin, std::move(callback));
+
+  api_request_helper_->Request("GET", url, "", "", true,
+                               std::move(internal_callback));
+}
+
+void AssetDiscoveryManager::OnFetchNFTsFromSimpleHash(
+    std::vector<mojom::BlockchainTokenPtr> nfts_so_far,
+    mojom::CoinType coin,
+    FetchNFTsFromSimpleHashCallback callback,
+    APIRequestResult api_request_result) {
+  if (!api_request_result.Is2XXResponseCode()) {
+    std::move(callback).Run(std::move(nfts_so_far));
+    return;
+  }
+
+  // Invalid JSON becomes an empty string after sanitization
+  if (api_request_result.body().empty()) {
+    std::move(callback).Run(std::move(nfts_so_far));
+    return;
+  }
+
+  absl::optional<std::pair<GURL, std::vector<mojom::BlockchainTokenPtr>>>
+      result = ParseNFTsFromSimpleHash(api_request_result.value_body(), coin);
+  if (!result) {
+    std::move(callback).Run(std::move(nfts_so_far));
+    return;
+  }
+
+  for (auto& token : result.value().second) {
+    nfts_so_far.push_back(std::move(token));
+  }
+
+  // If there is a next page, fetch it
+  if (result.value().first.is_valid()) {
+    auto internal_callback =
+        base::BindOnce(&AssetDiscoveryManager::OnFetchNFTsFromSimpleHash,
+                       weak_ptr_factory_.GetWeakPtr(), std::move(nfts_so_far),
+                       coin, std::move(callback));
+    api_request_helper_->Request("GET", result.value().first, "", "", true,
+                                 std::move(internal_callback));
+    return;
+  }
+
+  // Otherwise, return the nfts_so_far
+  std::move(callback).Run(std::move(nfts_so_far));
+}
+
+absl::optional<std::pair<GURL, std::vector<mojom::BlockchainTokenPtr>>>
+AssetDiscoveryManager::ParseNFTsFromSimpleHash(const base::Value& json_value,
+                                               mojom::CoinType coin) {
+  // Only ETH and SOL NFTs are supported.
+  if (!(coin == mojom::CoinType::ETH || coin == mojom::CoinType::SOL)) {
+    return absl::nullopt;
+  }
+
+  if (!json_value.is_dict()) {
+    return absl::nullopt;
+  }
+
+  GURL nextURL;
+  auto* next = json_value.FindStringKey("next");
+  if (next) {
+    nextURL = GURL(*next);
+  }
+
+  const base::Value* nfts = json_value.FindKey("nfts");
+  if (!nfts || !nfts->is_list()) {
+    return absl::nullopt;
+  }
+
+  const base::flat_map<std::string, std::string>& chain_ids =
+      FromSimpleHashChainId();
+  std::vector<mojom::BlockchainTokenPtr> nft_tokens;
+  for (const auto& nft : nfts->GetList()) {
+    auto token = mojom::BlockchainToken::New();
+
+    // Skip all tokens with a collection.spam_score > 0.
```

Comment:
@kdenhartog There's some documentation on their website - https://docs.simplehash.com/reference/spam-scores.  One of their stated goals is to identify "low-quality NFTs from undesirable airdrops", and they say their model "is generated through a combination of models, heuristics and drawing on external data sources", so it seems more involved then user reporting.  It seems to work fairly well based on my limited testing.

Reply:
Ok, I'm good with keeping this as is and just relying on their spam scoring system. Would it be too difficult to add a setting for this as well to allow the user to configure the spam score and keep the default 0? 

I'd be fine to leave this out of scope and do separately. Main reason I'm thinking this would be useful is I can already see users complaining because we're not automatically discovering their NFT because the collection has a relatively small spam score, but still greater than 0.

Reply:
Right, I think we will iterate on this more in the future.  Either by adjusting the spam score as you say, or allowing users to see 'NFTs marked as spam'. For now we should be conservative (there is a lot of spam, esp on cheaper networks).  Users always can fallback to adding their NFTs manually.

Reply:
Sounds good - I'm good with that. Going to resolve this one